### PR TITLE
Add step builtins and step HLSL function to DirectX and SPIR-V backend

### LINKED ISSUE #99157: Implement the `step` HLSL Function
- [ ] Implement `step` clang builtin,
- [ ] Link `step` clang builtin with `hlsl_intrinsics.h`
- [ ] Add sema checks for `step` to `CheckHLSLBuiltinFunctionCall` in `SemaChecking.cpp`
- [ ] Add codegen for `step` to `EmitHLSLBuiltinExpr` in `CGBuiltin.cpp`
- [ ] Add codegen tests to `clang/test/CodeGenHLSL/builtins/step.hlsl`
- [ ] Add sema tests to `clang/test/SemaHLSL/BuiltIns/step-errors.hlsl`
- [ ] Create the `int_spv_step` intrinsic in `IntrinsicsSPIRV.td`
- [ ] In SPIRVInstructionSelector.cpp create the `step` lowering and map  it to `int_spv_step` in `SPIRVInstructionSelector::selectIntrinsic`.
- [ ] Create SPIR-V backend test case in `llvm/test/CodeGen/SPIRV/hlsl-intrinsics/step.ll`

## DirectX

There were no DXIL opcodes found for `step`.

## SPIR-V

# Step:

## Description:
**Step**  
  
Result is 0.0 if *x* \< *edge*; otherwise result is 1.0.  
  
The operands must all be a scalar or vector whose component type is
floating-point.  
  
*Result Type* and the type of all operands must be the same type.
Results are computed per component.

<table>
<colgroup>
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 20%" />
</colgroup>
<thead>
<tr>
<th>Number</th>
<th>Operand 1</th>
<th>Operand 2</th>
<th>Operand 3</th>
<th>Operand 4</th>
</tr>
</thead>
<tbody>
<tr>
<td class="tableblock halign-left valign-top"><p>48</p></td>
<td
class="tableblock halign-left valign-top"><p><em>&lt;id&gt;</em><br />
<em>edge</em></p></td>
<td
class="tableblock halign-left valign-top"><p><em>&lt;id&gt;</em><br />
<em>x</em></p></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>

### Example 1
```hlsl
//dxc step_test.hlsl -T lib_6_8 -enable-16bit-types -O0

export float4 fn(float4 p1, float4 p2) {
    return step(p1, p2);
}
```
## HLSL:

Compares two values, returning 0 or 1 based on which value is greater.

| *ret* step(*y*, *x*) |
|----------------------|

 

## Parameters

| Item                                                   | Description                                                   |
|--------------------------------------------------------|---------------------------------------------------------------|
| <span id="y"></span><span id="Y"></span>*y*<br/> | \[in\] The first floating-point value to compare.<br/>  |
| <span id="x"></span><span id="X"></span>*x*<br/> | \[in\] The second floating-point value to compare.<br/> |

 

## Return Value

1 if the *x* parameter is greater than or equal to the *y* parameter; otherwise, 0.

## Remarks

This function uses the following formula: (*x* >= *y*) ? 1 : 0. The function returns either 0 or 1 depending on whether the *x* parameter is greater than the *y* parameter. To compute a smooth interpolation between 0 and 1, use the [**smoothstep**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-smoothstep.md) HLSL intrinsic function.

## Type Description

| Name  | [**Template Type**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md)                                                  | [**Component Type**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md) | Size                           |
|-------|----------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------|--------------------------------|
| *y*   | [**scalar**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md), **vector**, or **matrix** | [**float**](/windows/desktop/WinProg/windows-data-types)                        | any                            |
| *x*   | same as input *y*                                                                                              | [**float**](/windows/desktop/WinProg/windows-data-types)                        | same dimension(s) as input *y* |
| *ret* | same as input *y*                                                                                              | [**float**](/windows/desktop/WinProg/windows-data-types)                        | same dimension(s) as input *y* |

 

## Minimum Shader Model

This function is supported in the following shader models.

| Shader Model                                                                       | Supported                   |
|------------------------------------------------------------------------------------|-----------------------------|
| [Shader Model 2 (DirectX HLSL)](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-sm2.md) and higher shader models | yes                         |
| [Shader Model 1 (DirectX HLSL)](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-sm1.md)                          | yes (vs\_1\_1 and ps\_1\_4) |

 

## See also

<dl> <dt>

[**Intrinsic Functions (DirectX HLSL)**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md)
</dt> </dl>

### PR DESCRIPTION
**Pull request by bob80905**
This PR adds the step intrinsic and an HLSL function that uses it.
The SPIRV backend is also implemented.

Used https://github.com/llvm/llvm-project/pull/102683 as a reference.
Fixes https://github.com/llvm/llvm-project/issues/99157

### CODE REVIEW

**Review by damyanp — APPROVED** [review]

**Review by python3kgae — APPROVED** [review]

File: clang/test/CodeGenHLSL/builtins/step.hlsl [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+// DXIL_NATIVE_HALF: call half @llvm.dx.step.f16(half
+// SPIR_NATIVE_HALF: call half @llvm.spv.step.f16(half
+// DXIL_NO_HALF: call float @llvm.dx.step.f32(float
+// SPIR_NO_HALF: call float @llvm.spv.step.f32(float
+// NATIVE_HALF: ret half
+// NO_HALF: ret float
+half test_step_half(half p0, half p1)
+{
+    return step(p0, p1);
+}
```
Comment (bogner) [inline clang/test/CodeGenHLSL/builtins/step.hlsl:27]: The in-tree builtin tests were updated recently to clean up a bunch of the differences between DXIL and SPIRV (See #105930). Could you please update this test to that newer style?

File: llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp [hunk 2]
```diff
@@ -320,6 +321,28 @@ static Value *expandPowIntrinsic(CallInst *Orig) {
...
+static Value *expandStepIntrinsic(CallInst *Orig) {
+
+  Value *X = Orig->getOperand(0);
+  Value *Y = Orig->getOperand(1);
+  Type *Ty = X->getType();
+  IRBuilder<> Builder(Orig);
+
+  Constant *one = ConstantFP::get(Ty->getScalarType(), 1.0);
+  Constant *zero = ConstantFP::get(Ty->getScalarType(), 0.0);
+  Value *cond = Builder.CreateFCmpOLT(Y, X);
```
Comment (bogner) [inline llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:333]: Style nit: Variables should start with capital letters (One, Zero, Cond)

File: llvm/test/CodeGen/DirectX/step.ll [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+declare float @llvm.dx.step.f32(float, float)
+declare <2 x float> @llvm.dx.step.v2f32(<2 x float>, <2 x float>)
+declare <3 x float> @llvm.dx.step.v3f32(<3 x float>, <3 x float>)
+declare <4 x float> @llvm.dx.step.v4f32(<4 x float>, <4 x float>)
+
+define noundef half @test_step_half(half noundef %p0, half noundef %p1) {
+entry:
+  ; CHECK: %0 = fcmp olt half %p1, %p0
+  ; CHECK: %1 = select i1 %0, half 0xH0000, half 0xH3C00
+  ; DOPCHECK: asdahg
```
Comment (bogner) [inline llvm/test/CodeGen/DirectX/step.ll:20]: Typo / something left over from debugging?

File: llvm/test/CodeGen/SPIRV/hlsl-intrinsics/step.ll [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  ; CHECK: %[[#]] = OpFunction %[[#vec4_float_32]] None %[[#]]
+  ; CHECK: %[[#arg0:]] = OpFunctionParameter %[[#vec4_float_32]]
+  ; CHECK: %[[#arg1:]] = OpFunctionParameter %[[#vec4_float_32]]
+  ; CHECK: %[[#]] = OpExtInst %[[#vec4_float_32]] %[[#op_ext_glsl]] Step %[[#arg0]] %[[#arg1]]
+  %hlsl.step = call <4 x float> @llvm.spv.step.v4f32(<4 x float> %a, <4 x float> %b)
+  ret <4 x float> %hlsl.step
+}
+
+declare <4 x half> @llvm.spv.step.v4f16(<4 x half>, <4 x half>)
+declare <4 x float> @llvm.spv.step.v4f32(<4 x float>, <4 x float>)
```
Comment (bogner) [inline llvm/test/CodeGen/SPIRV/hlsl-intrinsics/step.ll:33]: Please add a newline at the end of the file

File: clang/lib/Sema/SemaHLSL.cpp [hunk 5]
```diff
@@ -1490,6 +1490,18 @@ bool SemaHLSL::CheckBuiltinFunctionCall(unsigned BuiltinID, CallExpr *TheCall) {
     TheCall->setType(ArgTyA);
     break;
   }
+  case Builtin::BI__builtin_hlsl_step: {
+    if (CheckFloatOrHalfRepresentations(&SemaRef, TheCall))
+      return true;
+    if (SemaRef.checkArgCount(TheCall, 2))
```
Comment (farzonl) [inline clang/lib/Sema/SemaHLSL.cpp:1496]: NIT: we usually check the arg count first.

File: clang/test/SemaHLSL/BuiltIns/step-errors.hlsl [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+{
+  return __builtin_hlsl_step(p1, p1);
+  // expected-error@-1 {{passing 'int' to parameter of incompatible type 'float'}}
+}
+
+bool2 builtin_step_int2_to_float2_promotion(int2 p1)
+{
+  return __builtin_hlsl_step(p1, p1);
+  // expected-error@-1 {{passing 'int2' (aka 'vector<int, 2>') to parameter of incompatible type '__attribute__((__vector_size__(2 * sizeof(float)))) float' (vector of 2 'float' values)}}
+}
```
Comment (farzonl) [inline clang/test/SemaHLSL/BuiltIns/step-errors.hlsl:31]: add a new line

**Review by farzonl — APPROVED** [review]

**Pull request merged**